src/cpp/src/prompt_lookup/continuous_batching_for_prompt_lookup.cpp

```diff
@@ -0,0 +1,85 @@
+
+namespace ov::genai {
+
+std::map<uint64_t, ContinuousBatchingPipeline::ContinuousBatchingForPromptLookupImpl::SequenceLen>
+ContinuousBatchingPipeline::ContinuousBatchingForPromptLookupImpl::get_generated_request_len() {
+    std::map<uint64_t, ContinuousBatchingPipeline::ContinuousBatchingForPromptLookupImpl::SequenceLen> result;
+    for (const auto& request : m_requests) {
+        const auto request_id = request->get_request_id();
+        auto validation_len = request->get_num_tokens_to_validate();
+        auto generated_len = request->get_num_processed_tokens() - request->get_prompt_len() + 1;
+        result.insert({ request_id, { generated_len, validation_len } });
+    }
+    return result;
+}
+
+TokenIds ContinuousBatchingPipeline::ContinuousBatchingForPromptLookupImpl::generate_candidates(const TokenIds& input_ids, size_t num_pred_tokens) {
+    if (num_pred_tokens == 0) {
+        return std::vector<int64_t>{};
+    }
+
+    const size_t input_length = input_ids.size();
+
+    for (int32_t ngram_size = m_max_ngram_size; ngram_size > 0; ngram_size--) {
+        // extract last ngram_size tokens as search ngram
+        std::vector<int64_t> ngram = std::vector<int64_t>{input_ids.cend() - ngram_size, input_ids.cend()};
+
+        // find ngram match in input_ids
+        size_t ngram_i = 0;
+        for (size_t input_i = 0; input_i < input_length - ngram_size; input_i++) {
+            if (ngram[ngram_i] != input_ids[input_i]) {
+                ngram_i = 0;
+                continue;
+            }
+
+            ngram_i++;
+
+            if (ngram_i < ngram_size) {
+                continue;
+            }
+
+            // match found with the end at input_i
+            size_t avaliable_num_pred = std::min(input_length - (input_i + 1), num_pred_tokens);
+
+            // return candidates with length of avaliable_num_pred
+            return std::vector<int64_t>{input_ids.cbegin() + input_i + 1,
+                                        input_ids.cbegin() + input_i + 1 + avaliable_num_pred};
+        }
+    }
+
+    return std::vector<int64_t>{};
+}
+
+void ContinuousBatchingPipeline::ContinuousBatchingForPromptLookupImpl::generate_candidates() {
+    for (auto& request : m_requests) {
+        const auto prompt = request->get_prompt_ids();
+        size_t max_validation_len = 0;
+        for (auto& running_sequence : request->get_running_sequences()) {
+            const auto generated_tokens = running_sequence->get_generated_ids();
+            TokenIds full_input_ids = prompt;
+            full_input_ids.insert(full_input_ids.end(), generated_tokens.begin(), generated_tokens.end());
+
+            size_t min_num_assistant_tokens = 0;
+            {
+                const auto generated_len = running_sequence->get_generated_len();
+                const auto sampling_params = request->get_sampling_parameters();
+                const auto left_generated_len = sampling_params.max_new_tokens - generated_len - 1;
```

Comment:
1. what if `max_new_tokens` is not specified?
2. there is an alternative way to stop generation by length - via https://github.com/openvinotoolkit/openvino.genai/blob/7d8912ff9df9bcfacf0044d108963cb7618bff69/src/cpp/include/openvino/genai/generation_config.hpp#L87

Reply:
Done